core/test/com/google/inject/ProvisionExceptionsTest.java

```diff
@@ -152,20 +170,70 @@ Exploder exploder() throws IOException {
                 return Explosion.createChecked();
               }
             });
-    try {
-      injector.getInstance(Tracer.class);
-      fail();
-    } catch (ProvisionException pe) {
-      pe.printStackTrace();
-      // Make sure our initial error message gives the user exception.
-      Asserts.assertContains(
-          pe.getMessage(), "1) [Guice/ErrorInCustomProvider]: IOException: boom!");
-      assertEquals(1, pe.getErrorMessages().size());
-      assertEquals(IOException.class, pe.getCause().getClass());
-      assertEquals(IOException.class, Messages.getOnlyCause(pe.getErrorMessages()).getClass());
+    var pe =
+        assertThrows(
+            ProvisionException.class,
+            () -> {
+              injector.getInstance(Tracer.class);
+            });
+    pe.printStackTrace();
+    // Make sure our initial error message gives the user exception.
+    Asserts.assertContains(pe.getMessage(), "1) [Guice/ErrorInCustomProvider]: IOException: boom!");
+    assertEquals(1, pe.getErrorMessages().size());
+    assertEquals(IOException.class, pe.getCause().getClass());
+    assertEquals(IOException.class, Messages.getOnlyCause(pe.getErrorMessages()).getClass());
+  }
+
+  static final class Errorer {
+    @Inject
+    Errorer() {
+      throw new OutOfMemoryError("uh oh");
+    }
+  }
+  static final class MethodErrorer {
+    @Inject MethodErrorer() {}
+    @Inject void injectErrorer() {
+      throw new OutOfMemoryError("uh oh");
     }
   }
 
+  // Demonstrate that different bindings do and do not wrap Errors thrown.
+  // Constructor and method injection does since the reflection API does it for us via
+  // InvocationTargetException.
+  // Provider instances do not since we have our own catch clauses and only catch RuntimeException
+  @Test
+  public void testErrorPropagation() {
```

Comment:
The exception management changes fixed most issues, but there's still one diff that's surfacing. The code follows a pattern of 

```
Provider<A> aProvider = { throw SomeError(); }
class BProvider implements Provider<B> {
  @Inject BProvider(A a) { .. }
}
// setup module binding A to aProvider instance
// setup module binding B to BProvider.class
// ask injection to get an instance of B
// expected SomeError, currently see ProvisionException
```

Reply:
Fun! This one was a little trickier but not too bad.  Basically we were placing the catch block around both the constructor binding and the constructor parameter bindings and it really needs to be just around the constructor invocation.

Found a similar issue with method injection

Finally found an inconsistency between constructor and provides injection in how exception sources are reported.  Documented that with a test.

Reply:
Oh also staring at this revealed an issue with the reflective fallbacks as well (we weren't always unwrapping InvocationTargetException), so that is fixed as well.